internal/pkg/agent/application/upgrade/marker_watcher.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type MarkerWatcher interface {
+	Watch() <-chan UpdateMarker
+	Errors() <-chan error
+	Run(ctx context.Context) error
+	Close() error
+}
+
+type MarkerFileWatcher struct {
+	watcher        *fsnotify.Watcher
+	markerFilePath string
+
+	logger *logger.Logger
+
+	updateCh chan UpdateMarker
+	errCh    chan error
+}
+
+func newMarkerFileWatcher(upgradeMarkerFilePath string, logger *logger.Logger) (MarkerWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create upgrade marker watcher: %w", err)
+	} // Watch the upgrade marker file's directory, not the file itself, so we
+
+	// notice the file even if it's deleted and recreated.
+	upgradeMarkerDirPath := filepath.Dir(upgradeMarkerFilePath)
+	err = watcher.Add(upgradeMarkerDirPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set watch on upgrade marker's directory [%s]: %w", upgradeMarkerDirPath, err)
+	}
+
+	return &MarkerFileWatcher{
+		watcher:        watcher,
+		markerFilePath: upgradeMarkerFilePath,
+		logger:         logger,
+		updateCh:       make(chan UpdateMarker),
+		errCh:          make(chan error),
+	}, nil
+}
+
+func (mfw *MarkerFileWatcher) Watch() <-chan UpdateMarker {
+	return mfw.updateCh
+}
+
+func (mfw *MarkerFileWatcher) Errors() <-chan error {
+	return mfw.errCh
+}
+
+func (mfw *MarkerFileWatcher) Run(ctx context.Context) error {
+	// Handle watching
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err, ok := <-mfw.watcher.Errors:
+				mfw.logger.Debug("after there are watch errors")
```

Comment:
[Suggestion]
Could you add a wee more context to be more clear after _what_ there are watch errors?

Reply:
Actually this is a holdover from some debugging I was doing.  This line should just be deleted.

Reply:
Deleted as part of the refactoring done in https://github.com/elastic/elastic-agent/commit/83577da72817b8c9817f1cd3f7b467825a9654f8.